### DISCUSSION TITLE
컴포넌트간에 공유하는 유틸과 타입에 대한 관리

### DISCUSSION THREAD

**Discussion by otterp012** [over-ui/unstyled#14] [General]
## 상황
`polymoriphic` 타입의 경우, **모든 컴포넌트**가 이 타입을 공유해야 합니다.
다만, 현재의 환경설정이 각 컴포넌트 별로 개별적인 패키지로 형성되어 있어 이 타입을 공유하기 위해서는 새로운 패키지를 작성해야 합니다.
이러한 이유로, `poly` 라는 패키지로 작성하고 있었는데 `polymoriphic` 타입 외에도 **모든 컴포넌트가 대부분 사용할 가능성이 높은 `utils`와
`hooks` 들을 별개의 패키지로 모아 작성하게 된다면 관리해야할 패키지의 개수가 너무 많아진다는 생각이 들었습니다.**

## 제안사항
이러한 패키지들을 한군데에 모아 `shared` 또는 `core`라는 이름으로 하나의 패키지로 만드는 방법은 어떨까요?

## 예시
`composeEvent` 라는 유틸 함수는, 모든 패키지에서 사용할 가능성이 높습니다.
```
function composeEvent<E>(
  defaultEvent: (event: E) => void,
  newEvent?: (event: E) => void,
) {
  return (event: E) => {
    defaultEvent?.(event);
    newEvent?.(event);
  };
```
이러한 유틸함수들을 모아, `core` 라는 하나의 패키지에 담아 사용합니다.

```
// core package
- poly types
- composeEvent
- 이외에 공유될 가능성이 높은 함수, 훅들
```

### 예상되는 효과
- 모든 패키지들이 사용할 가능성 높은 유틸, 훅, 타입을 한군데에서 관리할 수 있으므로 관리가 용이합니다.

### 예상되는 단점
- `core`라는 패키지가 중구난방 형식으로 각종 조그마한 유틸을 모아두는 유틸모음집이 될 가능성이 높습니다.
   추후에 `core` 라는 패키지가 가지는 의미가 퇴색될 가능성이 있습니다. (언제나 이 부분에 신경을 써야합니다.)
- `core`의 크기가 커질수록, 다른 컴포넌트의 모듈 사이즈도 커지게 됩니다.

### 추가되는 논의점
- `core` 패키지를 이용하게 된다면 `core`패키지에 추가하게 되는 기준을 정해야 합니다.
- 예를 들자면, **적어도 3개 이상의 컴포넌트에서 공유하고 있다면 `core`로 추가할 것** 이라는 기준이 필요합니다.

### 관련 이슈
#10 (#16 )

**Comment by lv0314** [comment]
좋은 제안입니다
안그래도 컴포넌트를 만들다보니, 여러 컴포넌트에서 공통적으로 사용되는 유틸 함수와 훅을 어떻게 관리해야할까 고민하던 참이었습니다
다만 언급해 주신 것 처럼 core의 크기가 커지면서 생기는 단점들이 있기 때문에 조금 신중하게 접근해야 할 것 같습니다
기준을 어떻게 잡느냐가 관건일 것 같은데, 횟수로 기준을 잡으면 조금 애매할 것 같다는 생각입니다
저희가 참고를 한 radix에서 useCallbackRef, useControllableState, useComposedRef 등 많이 사용되는 훅들도 core에 들어가지 못하는걸로 보아 정말 웬만한 컴포넌트에서 공통적으로 쓰이는 것들(예시의 poly, composeEvent 등)만이 core에 들어가야 할 것 같습니다
기준 잡기가 굉장히 까다롭네요...
